url10fb6bd43e_guwere_P2PSearch_tgz-pJ8-MainJ8/src/Controller.java

```diff
@@ -0,0 +1,577 @@
+            		}else{
+            			
+            		}             	           				           		 
+            	}           	
+                break;
+                //internal housekeeping. send out pings to stale contacts
+            case REFRESH_ROUTING_TABLE:
+            	//System.out.println("will handle update routing table message");
+               	for (Map.Entry<Integer, Contact> entry : routingTable.entrySet()){
+               		final Contact contact = entry.getValue();
+               		long currTime = System.currentTimeMillis();
+               		if(currTime - contact.getTimestamp() > NetworkProperties.staleTime*1000){	               		
+	        			message = getPingJson(contact);
+	        			address = new InetSocketAddress(contact.getIp(),NetworkProperties.nPort);           				
+				        sendQueue.put(jsonToDatagram(message,address));
+				        Timer pingTimer = new Timer();
+				        pingTimer.schedule(new TimerTask() {
+							//if timer not cancelled before time expires then contact removed
+							public void run() {
+								routingTable.remove(contact.getId());							
+							}
+						}, NetworkProperties.pingWait*1000);
+				        pingTimers.put(contact.getId(), pingTimer);
+	        		}
+               	}
+            	break;
+        }
+        }catch(InterruptedException e){
+          System.err.println("Java!! You ain't serious, are you? Anyway, couldn't put message on sendQ in Controller");
+          e.printStackTrace();
+        }
+    }
+
+
+
+	@SuppressWarnings("unchecked")
+    public void sendJoin(InetSocketAddress bootstrap_node,
+                         int nodeHash,
+                         int bootHash){
+        nodeID = nodeHash;
+        bootstrapNode = bootstrap_node;
+        if(nodeHash == bootHash) return;
+
+        JSONObject sendJSON = new JSONObject();
+        sendJSON.put("type",MessageType.JOINING_NETWORK_SIMPLIFIED.name());
+        sendJSON.put("node_id",Integer.toString(nodeHash));
+        sendJSON.put("target_id",Integer.toString(bootHash));
+        sendJSON.put("ip_address",NetworkProperties.nAddress.toString());
+        try {
+            sendQueue.put(jsonToDatagram(sendJSON,bootstrap_node));
+        } catch (InterruptedException e) {
+            System.err.println("Could not put Datagram on sendQueue in Controller");
+            e.printStackTrace();
+        }
+
+    }
+	@SuppressWarnings("unchecked")
+	public void leaveNetwork(long network_id) {
+        //tell of me contact I am leaving
+		for (Map.Entry<Integer, Contact> entry : routingTable.entrySet()){
+			JSONObject message = new JSONObject();
+			message.put("type", "LEAVE_NETWORK");
+			message.put("node_id", String.valueOf(nodeID));
+			InetSocketAddress address = new InetSocketAddress(entry.getValue().getIp(),NetworkProperties.nPort); 
+			try {
+	            sendQueue.put(jsonToDatagram(message,address));
+	        } catch (InterruptedException e) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: InterruptedException is ignored. This can delay thread shutdown and clear the thread’s interrupt status. Only code that implements a thread’s interruption policy can swallow an interruption request.

**Fix**: Rethrow the InterruptedException or reinterrupt the current thread using *Thread.currentThread().interrupt()* so that higher-level interrupt handlers can function correctly.
 If you are wrapping the InterruptedException inside a RuntimeException, call *Thread.currentThread().interrupt()* before throwing the RuntimeException.

Learn more about [interrupts](https://docs.oracle.com/javase/tutorial/essential/concurrency/interrupt.html) and [dealing with InterruptedException](https://www.ibm.com/developerworks/java/library/j-jtp05236/index.html?ca=drs-#N10187)